src/mds/Mantle.cc

```diff
@@ -16,144 +16,110 @@
-    return -EINVAL;
-  }
-
-  return 0;
-}
-
-int Mantle::balance(const string &script,
-                    mds_rank_t whoami,
-                    const vector < map<string, double> > &metrics,
-                    map<mds_rank_t,double> &my_targets)
-{
-  if (start() != 0)
-    return -ENOEXEC;
-
   /* tell the balancer which mds is making the decision */
-  lua_pushinteger(L, int(whoami));
-  lua_setfield(L, -2, "whoami");
+  lua_pushinteger(L, (lua_Integer)whoami);
+  lua_setglobal(L, "whoami");
 
   /* global mds metrics to hold all dictionaries */
   lua_newtable(L);
 
   /* push name of mds (i) and its metrics onto Lua stack */
-  for (unsigned i=0; i < metrics.size(); i++) {
-    lua_pushinteger(L, i);
+  for (size_t i=0; i < metrics.size(); i++) {
     lua_newtable(L);
 
     /* push values into this mds's table; setfield assigns key/pops val */
-    for (map<string, double>::const_iterator it = metrics[i].begin();
-         it != metrics[i].end();
-         ++it) {
-      lua_pushnumber(L, it->second);
-      lua_setfield(L, -2, it->first.c_str());
+    for (const auto &it : metrics[i]) {
+      lua_pushnumber(L, it.second);
+      lua_setfield(L, -2, it.first.c_str());
     }
 
     /* in global mds table at stack[-3], set k=stack[-1] to v=stack[-2] */
-    lua_rawset(L, -3);
+    lua_seti(L, -2, i);
   }
 
   /* set the name of the global mds table */
   lua_setglobal(L, "mds");
 
-  int ret = execute(script);
-  if (ret != 0) {
-    lua_close(L);
-    return ret;
+  assert(lua_gettop(L) == 1);
+  if (lua_pcall(L, 0, 1, 0) != LUA_OK) {
+    mantle_dout(0) << "WARNING: mantle could not execute script: "
+            << lua_tostring(L, -1) << mantle_dendl;
+    return -EINVAL;
   }
 
   /* parse response by iterating over Lua stack */
   if (lua_istable(L, -1) == 0) {
-    dout(0) << "WARNING: mantle script returned a malformed response" << dendl;
-    lua_close(L);
+    mantle_dout(0) << "WARNING: mantle script returned a malformed response" << mantle_dendl;
     return -EINVAL;
   }
 
   /* fill in return value */
-  mds_rank_t it = mds_rank_t(0);
   for (lua_pushnil(L); lua_next(L, -2); lua_pop(L, 1)) {
-    if (!lua_isnumber(L, -1)) {
-      dout(0) << "WARNING: mantle script returned a malformed response" << dendl;
-      lua_close(L);
+    if (!lua_isinteger(L, -2) || !lua_isnumber(L, -1)) {
+      mantle_dout(0) << "WARNING: mantle script returned a malformed response" << mantle_dendl;
       return -EINVAL;
     }
-    my_targets[it] = (lua_tonumber(L, -1));
-    it++;
+    mds_rank_t rank = mds_rank_t(lua_tointeger(L, -2));
+    my_targets[rank] = lua_tonumber(L, -1);
   }
 
-  lua_close(L);
   return 0;
 }
+
+Mantle::Mantle (void)
+{
+  /* build lua vm state */
+  L = luaL_newstate();
+  if (!L) {
+    mantle_dout(0) << "WARNING: mantle could not load Lua state" << mantle_dendl;
+    throw std::bad_alloc();
+  }
+
+  /* balancer policies can use basic Lua functions */
+  luaL_openlibs(L);
```

Comment:
Do we want to open all libraries? I think at one of the CDMs I was told that that could be dangerous (we don't want the OS modules, for example). `lua_openbase()` might be safer?